js/source/source.js

```diff
@@ -30,23 +35,53 @@ exports.create = function(id, source, dispatcher) {
     return source;
 };
 
-exports.getType = function (name) {
+Source.getType = function (name) {
     return sourceTypes[name];
 };
 
-exports.setType = function (name, type) {
+Source.setType = function (name, type) {
     sourceTypes[name] = type;
 };
 
 /**
+ * Returns the names of any registered non-core source types.
+ * @private
+ */
+Source.getCustomTypeNames = function () {
+    return Object.keys(sourceTypes).filter(function (type) {
+        return coreTypes.indexOf(type) < 0;
+    });
+};
+
+/**
+ * Adds a [custom source type](#Custom Sources), making it available for use with
+ * {@link Map#addSource}.
+ * @private
+ * @param {string} name The name of the source type; source definition objects use this name in the `{type: ...}` field.
+ * @param {Function} SourceType A {@link Source} constructor.
+ */
+Source.addType = function (name, SourceType) {
+    if (Source.getType(name)) {
+        throw new Error('A source type named ' + name + ' already exists.');
+    }
+
+    Source.setType(name, SourceType);
+
+    // an internal event, used to notify Style instances that there is a new
+    // custom source type.
+    Source.fire('_add', { name: name });
```

Comment:
This could be simplified a little bit now that we have a global worker pool. It could be simplified even more once we implement #3034 😄